src/providers/auth.ts

```diff
@@ -1,57 +1,300 @@
+  }
+
+
   private _emitAuthData(authData: FirebaseAuthDataAllProviders): void {
     if (authData == null) {
       this.next(null);
     } else {
-      let {uid, provider, github} = authData;
-      let authState: FirebaseAuthState = {uid, provider: null};
-      switch (provider) {
-        case "github":
-          authState.github = github;
-          authState.provider = AuthProviders.Github;
-          break;
-        default:
-          throw new Error(`Unsupported firebase auth provider ${provider}`);
+      this.next(this._authDataToAuthState(authData));
+    }
+  }
+
+  private _authWithCustomToken(token: string, options?: any): Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      this._fbRef.authWithCustomToken(token, this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  private _authAnonymously(options?: any): Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      this._fbRef.authAnonymously(this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  private _authWithPassword(credentials: FirebaseCredentials, options?: any)
+    : Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      this._fbRef.authWithPassword(credentials, this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  private _authWithOAuthPopup(provider: AuthProviders, options?: any): Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      this._fbRef.authWithOAuthPopup(this._providerToString(provider),
+                                     this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  /**
+   * Authenticates a Firebase client using a redirect-based OAuth flow
+   * NOTE: This promise will not be resolved if authentication is successful since the browser redirected.
+   * You should subscribe to the FirebaseAuth object to listen succesful login
+   */
+  private _authWithOAuthRedirect(provider: AuthProviders, options?: any): Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      this._fbRef.authWithOAuthRedirect(this._providerToString(provider),
+                                        this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  private _authWithOAuthToken(provider: AuthProviders, credentialsObj: OAuthCredentials, options?: any)
+  : Promise<FirebaseAuthState> {
+    let p = new Promise((res, rej) => {
+      let credentials = credentialsObj.hasOwnProperty('token') ? (<OAuth2Credentials> credentialsObj).token 
+                                                               : credentialsObj;
+      this._fbRef.authWithOAuthToken(this._providerToString(provider), credentials,
+                                     this._handleFirebaseCb(res, rej), options);
+    });
+
+    return p;
+  }
+
+  private _providerToString(provider: AuthProviders): string {
```

Comment:
Let's move away from enums and use string literal types. `provider: 'github' | 'twitter'`